hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/ReconSchemaManager.java

```diff
@@ -29,29 +30,128 @@
+ *
+ * This class handles the following scenarios:
+ *
+ * 1. Fresh Installation:
+ *    - No tables, including `schemaVersionTable`, exist.
+ *    - All tables are created with the latest schema version.
+ *
+ * 2. Upgrade from Older Version:
+ *    - All Existing tables (e.g., `UNHEALTHY_CONTAINERS`) are present, but `schemaVersionTable` is missing.
+ *    - Indicates an upgrade from a version without schema tracking which was introduced in version 2.0.
+ *    - The `schemaVersionTable` is created and all tables are upgraded to the latest version.
+ *
+ * 3. Upgrade with SchemaVersionTable:
+ *    - `schemaVersionTable` exists but is outdated.
+ *    - Migrations are applied to upgrade all tables to the latest schema.
+ *
+ * 4. Schema Already Up to Date:
+ *    - All tables and the schema version match the latest version; no action is needed.
  */
 public class ReconSchemaManager {
 
   private static final Logger LOG =
       LoggerFactory.getLogger(ReconSchemaManager.class);
   private Set<ReconSchemaDefinition> reconSchemaDefinitions = new HashSet<>();
+  private final ReconSchemaVersionTableManager schemaVersionTableManager;
 
   @Inject
-  public ReconSchemaManager(Set<ReconSchemaDefinition> reconSchemaDefinitions) {
+  public ReconSchemaManager(Set<ReconSchemaDefinition> reconSchemaDefinitions,
+                            ReconSchemaVersionTableManager schemaVersionTableManager) {
+    this.schemaVersionTableManager = schemaVersionTableManager;
     this.reconSchemaDefinitions.addAll(reconSchemaDefinitions);
   }
 
   @VisibleForTesting
-  public void createReconSchema() {
-    reconSchemaDefinitions.forEach(reconSchemaDefinition -> {
+  public void createAndUpgradeReconSchema() {
+
+    boolean isUpgrade = areOtherTablesExisting();
+
+    // Initialize all tables
+    initializeAllSchemas();
+
+    // Fetch current version from the SchemaVersionTable
+    String currentVersion = schemaVersionTableManager.getCurrentSchemaVersion();
+    String latestVersion = ReconConstants.LATEST_SCHEMA_VERSION;
+
+    // Handle cases where currentVersion is null, indicating schemaVersionTable just got created
+    if (currentVersion == null) {
+      if (isUpgrade) {
+        // Case 1: Upgrade from older version where schemaVersionTable was not present
+        LOG.info("Upgrade from older version detected. Setting current schema version to 1.0.");
+        currentVersion = "1.0"; // Set current version to the previous version before schemaVersionTable was introduced
+      } else {
+        // Case 2: Fresh install
+        LOG.info("Fresh installation detected. Setting schema version to latest.");
+        currentVersion = latestVersion;
+      }
+    }
+
+    // Upgrade schema if necessary
+    if (!currentVersion.equals(latestVersion)) {
+      upgradeAllSchemas(currentVersion, latestVersion);
+      // Update the schema version in the version table after migration
+      schemaVersionTableManager.updateSchemaVersion(latestVersion);
```

Comment:
This seems incorrect here, we are updating the schema version in schemaVersionTable even if we get some error while upgrading the schema of a required table ? How are we handling failure cases.